File: 0236-lowest-common-ancestor-of-a-binary-tree/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        // “For every node, check if p and q exist in its left subtree and right subtree.”
        // case 1
        // p in left subtree
        // q in right subtree
        if(root == null || root == p || root == q) {
            return root;
        }
        // case 2
        // both p and q in left subtree
        TreeNode left = lowestCommonAncestor(root.left, p, q);
        // case 3
        // both p and q in right subtree
        TreeNode right = lowestCommonAncestor(root.right, p, q);

        if(left == null) {
            return right;
        } else if(right == null) {
            return left;
        } else {
            return root;
        }
    }
}

File: 0236-lowest-common-ancestor-of-a-binary-tree/solution.swift
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public var val: Int
 *     public var left: TreeNode?
 *     public var right: TreeNode?
 *     public init(_ val: Int) {
 *         self.val = val
 *         self.left = nil
 *         self.right = nil
 *     }
 * }
 */

class Solution {
    func lowestCommonAncestor(_ root: TreeNode?, _ p: TreeNode?, _ q: TreeNode?) -> TreeNode? {
        guard let root = root else {
            return root
        }

        if root === p || root === q {
            return root
        }
        var left = lowestCommonAncestor(root.left, p, q)
        var right = lowestCommonAncestor(root.right, p, q)
        
        // if both are present, return root
        if left != nil && right != nil {
            return root
        }

        // if left is present, return left
        if left != nil {
            return left
        }

        // if right is present , return right
        if right != nil {
            return right
        }
        return nil
    }
}
